### DISCUSSION TITLE
Why are new RawContacts getting inserted with null account instead of custom account?

### DISCUSSION THREAD

**Discussion by marrale** [vestrel00/contacts-android#334] [Q&A] [answered]
To add an account to a contact I use following code:
```
const val ACCOUNT_NAME = "com.test"
const val ACCOUNT_TYPE = "Contacts"

fun toNewRawContact(): NewRawContact {
    val newRawContact = NewRawContact(
        sourceId = id,
        name = NewName(
            givenName = firstName,
            middleName = middleName,
            familyName = lastName
        ),
        account = Account(ACCOUNT_NAME, ACCOUNT_TYPE)
    )
    newRawContact.setDataAsReadOnly(true)
    return newRawContact
}


val rawContacts = myContactsList.map { c -> c.toNewRawContact() }
Contacts(inContext)
    .insert()
    .rawContacts(rawContacts)
    .commitInChunksWithContext()
```
The Contact's are getting inserted, but only on the Device itself and not to the specified Account.
I assume that I might forget to add something for the Contact's to be added to the right Account?

If I insert the Contact's via my Custom Code, everything get's inserted into the mentioned Account:
```
suspend fun addMultipleContacts(inContentResolver: ContentResolver, inContactDTOs: List<ContactDTO>){
    val contentProviderOperations = arrayListOf<ContentProviderOperation>()
    inContactDTOs.forEach { contactDTO ->
        val rawId = contentProviderOperations.size
        contentProviderOperations.add(
            ContentProviderOperation.newInsert(ContactsContract.RawContacts.CONTENT_URI)
                .withValue(ContactsContract.RawContacts.ACCOUNT_TYPE, ACCOUNT_TYPE)
                .withValue(ContactsContract.RawContacts.ACCOUNT_NAME, ACCOUNT_NAME)
                .withValue(ContactsContract.RawContacts.SOURCE_ID, contactDTO.id)
                .build()
        )

        val contentProviderOperationBuilder = ContentProviderOperation.newInsert(
            ContactsContract.Data.CONTENT_URI)
            .withValueBackReference(
                ContactsContract.Data.RAW_CONTACT_ID,
                rawId
            )
            .withValue(ContactsContract.Data.IS_READ_ONLY, 1) // Readonly == true
            .withValue(
                ContactsContract.Data.MIMETYPE,
                ContactsContract.CommonDataKinds.StructuredName.CONTENT_ITEM_TYPE
            )


        contentProviderOperationBuilder//
            .withValue(ContactsContract.CommonDataKinds.StructuredName.GIVEN_NAME, contactDTO.firstName)
            .withValue(ContactsContract.CommonDataKinds.StructuredName.FAMILY_NAME, contactDTO.lastName)
            .withValue(ContactsContract.CommonDataKinds.StructuredName.MIDDLE_NAME, contactDTO.middleName)

        contentProviderOperations.add(contentProviderOperationBuilder.build())
    }

    try {
        val a = inContentResolver.applyBatch(
            ContactsContract.AUTHORITY,
            contentProviderOperations
        )
    } catch (inE: Exception) {
        System.err.println("Can't insert LocalPhone Contacts!: ${inE.message} >>> ${inE.cause} >>> $inE")
    }
}
```

**Comment by vestrel00** [comment]
This might be regression caused by #325, which might only reproducible in the alpha build maybe.

Could you try to reproduce the issue in `0.3.1`?

**Reply by vestrel00** [reply]
I'll take a look later too when I get the time.

**Reply by marrale** [reply]
You are right, I have done the tests with the new 0.3.2-alpha1 version. Tomorrow I will be able to perform this test with the released version and give you feedbacks.

**Reply by vestrel00** [reply]
By the way, are you using a real account when inserting? By default, if the `Account` is not returned by the `android.accounts.AccountManager`, or more specifically this accounts query from the library (`contactsApi.accounts().query().find()`), then the account will be considered invalid and null account name and type will be used when inserting the `NewRawContact`.

You can disable this account validation in order to pass in any Account regardless of whether or not it is registered in the `AccountManager`,

```Kotlin
contactsApi
    .insert()
    .validateAccounts(false)
    ...
```

See https://vestrel00.github.io/contacts-android/basics/insert-contacts/#account-validation

Please let me know how it goes. Thanks!

**Reply by marrale** [reply]
Yes, it should be a real account that I create with this method:
```
    fun createAccount(inContext: Context) {
        System.err.println("createAccount called")
        try {
            val accountManager = AccountManager.get(inContext)
            val account = Account(ACCOUNT_NAME, ACCOUNT_TYPE)
            // It is successfully creating a Account, but it can't be shown for some reason...
            val success = accountManager.addAccountExplicitly(account, null, null)
            if (success) {
                System.err.println("Account created successfully")
            } else {
                System.err.println("Account creation failed. Look at previous logs to investigate (type: 'account' in Logcat without 'package:mine')")
                if (accountManager.accounts.isNotEmpty() && Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                    System.err.println("Accounts found!:")
                    accountManager.accounts.forEach {
                        if (it.name.contentEquals(ACCOUNT_NAME, true)) {
                            try {
                                val packagesAndVisibilityForAccount =
                                    accountManager.getPackagesAndVisibilityForAccount(it)
                                System.err.println("Account Name: ${it.name} >>> Type: ${it.type} >>> Visibility: $packagesAndVisibilityForAccount")
                                if (packagesAndVisibilityForAccount.isEmpty()) {
                                    System.err.println("Empty packagesAndVisibilityForAccount, try to set Visibility")
                                    accountManager.setAccountVisibility(
                                        it,
                                        inContext.packageName,
                                        AccountManager.VISIBILITY_VISIBLE
                                    )
                                    val newVisibility =
                                        accountManager.getPackagesAndVisibilityForAccount(it)
                                    System.err.println("New Visibility: $newVisibility")
                                }
                            } catch (ex: Exception) {
                                System.err.println("Account couldn't be updated to be visible: ${ex.localizedMessage}")
                            }
                        }
                    }
                }
            }
        } catch (ex: Exception) {
            System.err.println("Can't add Account exception: ${ex.localizedMessage}")
        }
    }
```
And this works when I use my custom code to insert the contact.
Tomorrow I will be able to test your suggestion to disable account validation.

**Reply by vestrel00** [reply, marked as answer]
I see that you are adding your own custom account via `AccountManager.addAccountExplicitly`. The Account validation performed in this library uses the `AccountsQuery` API to retrieve the visible Accounts that has a registered SyncAdapter for Contacts syncing; https://vestrel00.github.io/contacts-android/accounts/query-accounts/

So, the issue could be that your custom Account is not being returned by the `AccountManager` because of visibility issues AND/OR it does not have a matching registered SyncAdapter type.

Regardless, turning off the Account validation should (hopefully) allow you to insert RawContacts to the Account of your choice without issue 🤞

------------

I do have a question. Does your custom Account show up as an option when creating new contacts using other Contacts applications such as [Google Contacts](https://play.google.com/store/apps/details?id=com.google.android.contacts)?

**Reply by marrale** [reply]
@vestrel00 using the 0.3.2-alpha1 I have set "validateAccounts(false)" as suggested from you and now it works, the contacts belongs to the account. Our account has no registered SyncAdapter, that's probably the reason why the account was not added to contact after validation. Thanks for your help!
Yes, my custom account shows up as an option when creating a new contact using the google contacts app.

**Reply by vestrel00** [reply]
Glad everything works for you now! Awesome 😁

**Reply by marrale** [reply]
For Android 10 it works as described/whished. With Android 12 it not works, the contacts are added on the Device itself and not to the specified Account, but this has nothing to do with the library, but probably with the created account, missing of sync adapter!?. Why I have the same behavior as described also with my custom code. I will look deeper on this and post an update once the Solution is found.
